### ISSUE TITLE
When I modified shared config "File added to the system" event is occur

### ISSUE THREAD

**Issue by demantos** [wazuh/wazuh#2741]
```
|Wazuh version|Install type|Install method|Platform|
|3.7.2 | manager and agent | rpm package | CentOS Linux |
```

I have some problem about centralized configuration

I was modified shared config in /var/ossec/etc/shared/default/agent.conf and **`File added to the system`** event is occur. 

Then event detail is not about config file, but about system file (e.g /sbin/xfs_bmap).

Event count is 90000 (our agent is about 400ea)

![image](https://user-images.githubusercontent.com/5017649/53792072-2fa7f680-3f6e-11e9-9fbc-dc58f3420e75.png)

I think **`File added to the system`** event must occur about `/var/ossec/etc/shared/agent.conf` in agent, but event occur about system files.

Is this event is right? 

What can I do to prevent these events from happening?

**Comment by DaveVG1** [comment]
Hi,

First of all I apologise for the late response. I've been trying to reproduce this but I haven't managed to. Is it possible that you have the following in your syscheck configuration (in `ossec.conf` or `agent.conf`)?: 

`<directories check_all="yes">/bin,/sbin,/boot</directories>` 

It comes by default and it scans some of the folders that you're receiving alerts about. It's possible that changing the centralized configuration triggers the scans of those folders (specially if you hadn't them in your agent's `ossec.conf` in the first place, but you put them in `agent.conf`)

If you can't figure it out, I could check your full configuration (`ossec.conf` and `agent.conf`) to help you out.

Regards,
David

**Comment by demantos** [comment]
Yes, I have config of syscheck like this on wazuh-maanger and wazuh-agent.

```
<directories check_all="yes">/etc,/usr/bin,/usr/sbin</directories>
<directories check_all="yes">/bin,/sbin,/boot</directories>
```
 You mentioned `It's possible that changing the centralized configuration triggers the scans of those folders` but, **I have no change in above directories.**

I changed `/var/ossec/etc/shared/default/agent.conf` and very many `File added to the system` event is occur.
When change the file in `/var/ossec/etc/shared/default/`, same event is occur.

When change the centralized configuration, syscheck scans the directories specified in syscheck <directories> directives. When new file is founded, `File added to the system` event occur is right, but file specified in full_log is not new file. That file is already exists.

Regards

**Comment by DaveVG1** [comment]
I still trying to reproduce this without success. These are the steps that I've taken (just in case I'm not doing it right):

1. I have a 3.7.2 manager and agent connected. 
2. I restart the agent for the first time and wait until all scans have finished.
3. I change the `agent.conf` on the manager side.
4. Moments later, the agent is restarted and I get no unexpected alerts (not even the **File added to the system** alert from `agent.conf`)

Is it possible that you have a considerable number of files in the syscheck folders that you're monitoring? It could be flooding the agent event queue and generating the events that couldn't be generated in the previous scan after modifying the agent.conf which causes the agent to make another scan. To check if that's the case, you could try to configure a low frequency in syscheck (60, for example) and check if the same alerts that you're getting are generated after the indicated frequency (second scan).

If this doesn't help, please let me know what your configuration is (`ossec.conf` and `agent.con`) as it may help to figure out what's going on.

Regards.

**Comment by demantos** [comment]
here is my configuration. 
syscheck frequency is 1 hour.

**[wazuh-manager] agent.conf**
```
<agent_config>
  <!-- Shared agent configuration here -->
  <rootcheck>
    <ignore>/var/lib/docker/overlay</ignore>
    <ignore>/var/lib/docker/volumes</ignore>
  </rootcheck>
</agent_config>
```

**[wazuh-agent] ossec.conf**
```
<syscheck>
    <disabled>no</disabled>
    <!-- Frequency that syscheck is executed default every 1 hours -->
    <frequency>3600</frequency>
    <scan_on_start>yes</scan_on_start>
    <alert_new_files>yes</alert_new_files>
    <!-- Directories to check  (perform all possible verifications) -->
    <directories check_all="yes">/etc,/usr/bin,/usr/sbin</directories>
    <directories check_all="yes">/bin,/sbin,/boot</directories>
    <directories realtime="yes">/var/tmp,/dev/shm</directories>
    <directories check_all="yes">/etc/yum.repos.d/</directories>
    <!-- Files/directories to ignore -->
    <ignore>/etc/mtab</ignore>
    <ignore>/etc/hosts.deny</ignore>
    <ignore>/etc/mail/statistics</ignore>
    <ignore>/etc/random-seed</ignore>
    <ignore>/etc/random.seed</ignore>
    <ignore>/etc/adjtime</ignore>
    <ignore>/etc/httpd/logs</ignore>
    <ignore>/etc/utmpx</ignore>
    <ignore>/etc/wtmpx</ignore>
    <ignore>/etc/cups/certs</ignore>
    <ignore>/etc/dumpdates</ignore>
    <ignore>/etc/svc/volatile</ignore>
    <!-- Check the file, but never compute the diff -->
    <nodiff>/etc/ssl/private.key</nodiff>
    <skip_nfs>yes</skip_nfs>
</syscheck>

```

**Comment by vikman90** [comment]
Hi @demantos,

Sorry for the late reply. This issue got deprecated, let me explain you why. This problem is difficult to reproduce as it depends on many different conditions such as lost events.

The FIM engine from OSSEC to Wazuh 3.11 might produce false positives (as you're having) due to a double differential engine:

1. The agents hold an in-memory database containing the file state records that they sent to the manager. This database prevents agents from sending the same record to the manager every time they run a scan. Agents send the full database on startup; from that moment on, they report changes only.
2. Messages from the agent to the manager may be lost (client buffer full in the agent side, network congestion (using UDP), or manager stressed).
3. The manager compares every file record received against its database. If the file was not in the database, the manager produces an alert about a new file. If the file was in the database but any attribute changed, the manager produces an alert about file modified.

## Possible cause

This is what I think it's happening in your environment:

1. The agent starts up and sends the complete file list (including _/boot/*_ and _/sbin/*_). Some of the records get lost at any point between the agent's FIM engine and the manager's decoder.
2. The agent does not send those records again as they were been marked as reported.
3. You modified the _agent.conf_ file and that made the agent restart.
4. The agent starts up again and sends the complete file database by the second time. This time, the records that were lost reach the manager, and that makes it think those files are new. That's why the manager produces the alerts.

## How to fix that

This problem required a rework in the FIM engine. It was fixed by #3319 (version 3.12.0). Both differential engines were merged into the agent, and new the agent reports file creation/modification/deletion events instead of file states. Those events always produce an alert as they definitely mean a change in the file system. An event lost will never lead to a false positive.

I hope this helps you, and again, sorry for replying so late.
Closing this issue.
